### PR TITLE
Boost: Refactor module status to use option store

### LINKED ISSUE #29088: [Boost] Deal with non-standard `get_options` responses
Based on: https://github.com/Automattic/jpop-issues/issues/8091

A user has reported that they can't turn Boost features on or off smoothly. On investigation, it looks like they have a caching setup which preserves `bool` values which get written to their options in a non-standard way.

According to the WordPress codex, `bool(true)` becomes `'1'` when written to the options, then read again using `get_option`. However, depending on the caching solution used we can't guarantee this conversion.

We need to ensure that we don't compare options to `=== 1` without also allowing `=== true`.

### PR DESCRIPTION
**Pull request by haqadn**
Fixes #29088

## Proposed changes:
* Use options-store for module status in PHP
* Added a backward compatibility module to transfer module status to option-store

Switching the JS side is coming in a separate PR.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

None

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

* Install boost and try toggling all modules on and off
* Make sure the functionality of each module continues to work
* Install boost using [current trunk](https://jurassic.ninja/create/?jetpack-beta&branches.jetpack-boost=master&wp-debug-log), toggle module statuses there and update to this feature branch.
* Make sure your old state transferred smoothly.

### CODE REVIEW

File: projects/plugins/boost/compatibility/boost-1.8.0.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	}
+
+	return $old_value;
+}
+
+add_filter( 'default_option_jetpack_boost_ds_module_status_critical_css', 'jetpack_boost_180_option_fallback', 10, 2 );
+add_filter( 'default_option_jetpack_boost_ds_module_status_cloud_css', 'jetpack_boost_180_option_fallback', 10, 2 );
+add_filter( 'default_option_jetpack_boost_ds_module_status_render_blocking_js', 'jetpack_boost_180_option_fallback', 10, 2 );
+add_filter( 'default_option_jetpack_boost_ds_module_status_lazy_images', 'jetpack_boost_180_option_fallback', 10, 2 );
+add_filter( 'default_option_jetpack_boost_ds_module_status_image_guide', 'jetpack_boost_180_option_fallback', 10, 2 );
```
Comment (dilirity) [inline projects/plugins/boost/compatibility/boost-1.8.0.php:30]: Can we do the migration on plugin activation/update? Looking at when `default_option_*` is called, the way it's done might be okay.

File: projects/plugins/boost/app/lib/Status.php [hunk 2]
```diff
@@ -29,10 +34,16 @@ public function __construct( $slug ) {
...
+	public function get_ds_entry_name() {
+		return 'module_status_' . str_replace( '-', '_', $this->slug );
 	}
 
 	public function is_enabled() {
-		return '1' === get_option( $this->get_option_name( $this->slug ) );
+		return $this->ds->get();
 	}
 
 	public function update( $new_status ) {
```
Comment (dilirity) [inline projects/plugins/boost/app/lib/Status.php:49]: **Note I'm referring to the `update` method and not the change 😅** 

I think the casting in the method is no longer necessary.

**Review by dilirity — CHANGES_REQUESTED** [review]
Two minor things and it should be ready to 🚢

File: projects/plugins/boost/app/lib/Status.php (hunk 2 above)
Comment (haqadn) [reply projects/plugins/boost/app/lib/Status.php:49]: Casting to the ds has been removed but, outside of data-store usage, it's still necessary as it is coming from outside the data store in `update`.

File: projects/plugins/boost/compatibility/boost-1.8.0.php (hunk 1 above)
Comment (haqadn) [reply projects/plugins/boost/compatibility/boost-1.8.0.php:30]: We can, but changing it on access would move the data only once, just in time. This way we don't have to worry about corner cases for update/install, etc.

**Review by dilirity — APPROVED** [review]
:shipit:

**Pull request merged**